### PR TITLE
2024-12-13-ram-data-decay-research.md: initial draft

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+this document, this term will be used to represent operating system's view of
+memory as one(*) consecutive range of addresses.
+
+DRAM (Dynamic RAM) is a type of RAM that holds the data in capacitors, which
+loses its charge over time and must be periodically refreshed. This is in
+contrast to static RAM (SRAM) which doesn't require refreshing to hold its data.
+The capacitor, together with logic required to access its content, is called a
+DRAM cell. Multiple cells form a grid that is addressable by row and column,
+such grid is called a bank. Each bank (in most commonly available DRAM chips) is
+replied 4 or 8 times in parallel, so each row/column pair addresses 4 or 8 bits
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:43]:
```suggestion
replicated 4 or 8 times in parallel, so each row/column pair addresses 4 or 8 bits
```

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+of data per DRAM at once. On top of that, each DRAM can have multiple banks,
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:44]:
```suggestion
of data per DRAM at once. On top of that, each DIMM can have multiple banks,
```

i think that's what you meant?

Comment (mkopec) [reply blog/content/post/2024-12-13-ram-data-decay-research.md:44]: but DIMM is defined in the subsequent paragraph, so maybe "DRAM module"?

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 1 above)
Comment (mkopec) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/445f12165153db5d3d50208baad9c00cd20168fe

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+For all tests, air humidity was constant 33%. Temperatures were changing
+slightly, they are listed below next to the results.
+
+Results show number of changed bits, divided between transitions from 1 to 0 and
+from 0 to 1, as well as an average of the two.
+
+### Laptop - DDR5 SODIMM
+
+On this platform, we've struggled to get any preserved bits at all. We ended up
+disconnecting the battery and unplugging and, as quickly as humanely possible,
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:251]:
```suggestion
disconnecting the battery and unplugging and, as quickly as humanly possible,
```

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 4]
```diff
@@ -0,0 +1,357 @@
+---
+title: Research of RAM data remanence times
+abstract: "Dynamic RAM must be periodically refreshed to maintain its content.
+          JEDEC specifies how often this has to be done to ensure proper
+          operation. In this post, we will see for how long the data can be
+          recovered if it isn't."
+cover: /covers/image-file.png
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:7]: Needs a proper cover

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+On both platforms, impact of DIMM topology can be seen. In case of DDR5 SODIMM,
+distribution of changes in bits 0-31 roughly aligns with bits 32-63, especially
+with average values. Lower half has more 0 to 1 transitions, which in higher
+half is mirrored by 1 to 0 transitions. Border between DRAMs is also visible,
+although it isn't that clear, contrary to DDR4 DIMM. Results from the latter
+show a clear difference between groups of 8 consecutive bits, with average
+almost constant within the group.
+
+During a given run, some DRAMs start loosing their data faster than others. It
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:328]:
```suggestion
During a given run, some DRAMs start losing their data faster than others. It
```

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+looks like this isn't necessarily correlated with their maximal preservation
+time - take a look at group of bits 40-47 in DDR4. For delays in range 10-50
+seconds, it is always the one that looses the biggest part of its content, but
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:331]:
```suggestion
seconds, it is always the one that loses the biggest part of its content, but
```

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+For DDR4, chart of percentage of lost bits in function of time shows that the
+amount of lost data rises quickly with lower delays, but then starts to slow
+down when it gets close to 50%. The plot is also flattened on the left side,
+which may suggest that some constant, systematic error is added to time
+measurements. In fact, after applying back the power, it takes about 5 seconds
+before power LED gets lit on that platform.
+
+## Summary
+
+We've learned that RAM needs time to lost its content, and that this time varies
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:343]:
```suggestion
We've learned that RAM needs time to lose its content, and that this time varies
```

File: blog/content/post/2024-12-13-ram-data-decay-research.md [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+between both tested configuration. Unfortunately, we don't have enough data
```
Comment (mkopec) [inline blog/content/post/2024-12-13-ram-data-decay-research.md:344]:
```suggestion
between both tested configurations. Unfortunately, we don't have enough data
```

**Review by mkopec — CHANGES_REQUESTED** [review]

**Review by BeataZdunczyk — APPROVED** [review]

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 3 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 4 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 5 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 6 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 7 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

File: blog/content/post/2024-12-13-ram-data-decay-research.md (hunk 8 above)
Comment (krystian-hebel) [reply blog/content/post/2024-12-13-ram-data-decay-research.md]: https://github.com/3mdeb/news-and-ideas/pull/621/commits/478ab8793dac42e71f71d1b2f30aeb7b92d053c6

**Review by mkopec — APPROVED** [review]

**Pull request merged**